src/d3a/d3a_core/sim_results/kpi.py

```diff
@@ -0,0 +1,145 @@
+            self.total_self_consumption_wh += trade.offer.energy * 1000
+
+    def _accumulate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.producer_list and trade.buyer_origin in self.ess_list:
+            self.self_consumption_buffer_wh += trade.offer.energy * 1000
+
+    def _dissipate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.ess_list:
+            # self_consumption_buffer needs to be exhausted to total_self_consumption
+            # if sold to internal consumer
+            if trade.buyer_origin in self.consumer_list and self.self_consumption_buffer_wh > 0:
+                if (self.self_consumption_buffer_wh - trade.offer.energy * 1000) > 0:
+                    self.self_consumption_buffer_wh -= trade.offer.energy * 1000
+                    self.total_self_consumption_wh += trade.offer.energy * 1000
+                else:
+                    self.total_self_consumption_wh += self.self_consumption_buffer_wh
+                    self.self_consumption_buffer_wh = 0
+            # self_consumption_buffer needs to be exhausted if sold to any external agent
+            elif trade.buyer_origin not in [*self.ess_list, *self.consumer_list] and \
+                    self.self_consumption_buffer_wh > 0:
+                if (self.self_consumption_buffer_wh - trade.offer.energy * 1000) > 0:
+                    self.self_consumption_buffer_wh -= trade.offer.energy * 1000
+                else:
+                    self.self_consumption_buffer_wh = 0
+
+    def _accumulate_infinite_consumption(self, trade):
+        if trade.seller_origin in self.buffer_list and trade.buyer_origin in self.consumer_list:
+            self.total_self_consumption_wh += trade.offer.energy * 1000
+
+    def _dissipate_infinite_consumption(self, trade):
+        if trade.buyer_origin in self.buffer_list and trade.seller_origin in self.producer_list:
+            self.total_self_consumption_wh += trade.offer.energy * 1000
+
+    def _accumulate_energy_trace(self):
+        for c_area in self.consumer_area_list:
+            for market in c_area.past_markets:
+                for trade in market.trades:
+                    self._accumulate_self_consumption(trade)
+                    self._accumulate_self_production(trade)
+                    self._accumulate_self_consumption_buffer(trade)
+                    self._dissipate_self_consumption_buffer(trade)
+                    self._accumulate_infinite_consumption(trade)
+                    self._dissipate_infinite_consumption(trade)
+
+    def area_performance_indices(self, area):
+        self.producer_list = []
+        self.consumer_list = []
+        self.consumer_area_list = []
+        self.ess_list = []
+        self.buffer_list = []
+        self.total_energy_demanded_wh = 0
+        self.total_energy_produced_wh = 0
+        self.total_self_consumption_wh = 0
+        self.self_consumption_buffer_wh = 0
```

Comment:
Do you think you can make these variables NOT part of self?

Reply:
Then it has to be part of method parameter and then catch it as part of function return value.

@spyrostz Do you have any opinion on that?

Reply:
It annoys me too actually, because there is no clear way to see how these values are populated, and makes it harder to see where these variables are actually used. I would suggest to at least add a separate class that can hold these parameters, and then you can simply pass an object of this class to the methods. That way you will at least avoid storing these as members of the KPI class.